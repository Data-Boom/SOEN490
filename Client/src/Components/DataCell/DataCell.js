import React from 'react'
import Box from '@material-ui/core/Box'
import Button from '@material-ui/core/Button'

const DataCell = ({children}) => (
  <Box display="flex" marginLeft={"25%"} flexDirection="column" justify="center" alignItems="center" border={30} borderColor={"#2e3b52"} width={720} height={300} padding={10}>
    <img src={require('./uploadimage.png')} />
    <p>Choose a File or Drag/Drop here</p>
    <Button onClick = {DataCell.callAPI} variant="contained" t = {0.5}>Browse</Button>
  </Box>
)

function callAPI(){
  fetch("http://localhost:4000/dataupload")
}

export default DataCell