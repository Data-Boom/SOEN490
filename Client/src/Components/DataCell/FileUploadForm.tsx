import React, { useState } from 'react'

import { Alert } from '@material-ui/lab'
import Box from '@material-ui/core/Box'
import Button from '@material-ui/core/Button'
import { Container } from '@material-ui/core'
import { Snackbar } from '@material-ui/core'
import { SyntheticEvent } from 'react'
import { classStyles } from '../../appTheme'
import uploadimage from './uploadimage.png'

interface IProps {
  acceptFileFormat: string,
  onSubmit: (uploadedFile: File) => void,
  //return true if file isValid
  validateFile?: (uploadedFile: File) => boolean,
}

const fileInputId = 'fileInput'

export const FileUploadForm = (props: IProps) => {
  const { onSubmit, validateFile, acceptFileFormat } = props

  const [alertOpen, setAlertOpen] = useState(false)
  const [alertSuccess, setAlertSuccess] = useState(false)


  const handleSubmit = (event: SyntheticEvent<HTMLFormElement, Event>) => {
    const input: HTMLInputElement = event.target[fileInputId]
    const file: File = input.files[0]
    //if validate file was not provided will isValidFile = true
    const isValidFile = validateFile(file) //|| true
    console.log("vF is " + validateFile + "  and vF() is " + validateFile(file))
    if (isValidFile) {
      onSubmit(file)
      console.log('File successfully uploaded')
      setAlertSuccess(true)
    }
    else {
      setAlertOpen(true)
    }
  }

  return (
    <>
      {/* Snackbar is used to show an error on the screen when a wrong file type is selected for uploading */}
      <Snackbar open={alertOpen} autoHideDuration={3000} onClose={() => setAlertOpen(false)}>
        <Alert onClose={() => setAlertOpen(false)} severity="error">
          Failed to parse file
        </Alert>
      </Snackbar>

      <Snackbar open={alertSuccess} autoHideDuration={3000} onClose={() => setAlertSuccess(false)}>
        <Alert onClose={() => setAlertSuccess(false)} severity="success">
          File Successfully updated!!
        </Alert>
      </Snackbar>

      <Container>
        <Box className={classStyles().defaultBorder}>
          <img src={uploadimage} alt="Visual of clouds"></img>
          <form onSubmit={handleSubmit}>
            <div>
              <input type="file" accept={acceptFileFormat} id={fileInputId} />
              <Button variant="contained" type='submit' color='primary'> Upload this file! </Button>
            </div>
          </form>
        </Box>
      </Container>
    </>
  )
}