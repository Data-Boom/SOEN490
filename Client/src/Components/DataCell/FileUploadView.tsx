import DataCell from './DataCell'
import React from 'react'

const FileUploadView = () => {

  return (
    <> <h2> File Upload Page </h2>
      <DataCell />
    </>
  )
}

export default FileUploadView
