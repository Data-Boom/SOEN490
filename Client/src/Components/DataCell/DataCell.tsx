import Box from '@material-ui/core/Box'
import Button from '@material-ui/core/Button'
import Download from '@axetroy/react-download'
import { FileUploadForm } from './FileUploadForm'
import React from 'react'
import { rm } from "../../Assets/readMeMessage"

const fileFormat = 'application/json'
export default function DataCell() {

  const validateJson = (file: File) => {
    return file && file.type !== fileFormat
  }

  /**
     * Upon submission, the JSON file is extracted from the event and must be appended to formData
     * to be sent with API request.
     */
  const handleSubmit = async (jsonFile: File) => {
    const formData = new FormData()
    formData.append('file', jsonFile)

    const options = {
      method: 'POST',
      body: formData,
    }

    try {
      await fetch('http://localhost:4000/api/v1/dataExtract', options)
        .then(resp => resp.json())
        .then(result => {
        })
    }
    catch (err) {
      //todo add error handling
    }
  }

  return (
    <>
      <FileUploadForm
        onSubmit={handleSubmit}
        validateFile={validateJson}
        acceptFileFormat={fileFormat}
      />
      <Box p={4}>
        <div>
          {/**for downloading sample empty json file*/}
          <Download file="emptyJsonDataset.json" content={JSON.stringify("../../Assets/emptyJSFile.json", null, 2)}>
            <Button type="submit" variant="contained"> Download Sample JSON file </Button>
          </Download>
        </div>
        {/**for downnloading instructions readMe for users */}
        <div>
          <Download file="readMe.txt" content={rm}>
            <a href="http://localhost:3000/#/uploadFile"> Download JSON file submission instructions </a>
          </Download>
        </div>
      </Box>
    </>
  )
}
