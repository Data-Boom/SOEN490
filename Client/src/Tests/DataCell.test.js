import React from 'react';
import renderer from "react-test-renderer";
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import DataCell from '../Components/DataCell/DataCell'

configure({ adapter: new Adapter() });

describe('DataCell Component', () => {

    it("should match snapshot", () => {
        const tree = renderer.create(<DataCell />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render a file input field', () => {
        const wrapper = mount(<DataCell />)
        expect(wrapper.find('input').exists()).toBeTruthy();
    });

    /*it('calls the handleSubmit function on button click.', () => {
        // Currently fails, most likely because handleSubmit is an async function.
        const spy = jest.spyOn(console, 'log');
        const wrapper = mount(<DataCell />);

        const form = wrapper.find('#formSubmit');
        expect(form.exists()).toEqual(true);
        form.simulate('click');
        expect(spy).toBeCalledWith('handleSubmit Test');

    });*/
});

