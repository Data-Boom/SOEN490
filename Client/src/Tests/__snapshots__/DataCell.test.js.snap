// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataCell Component should match snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    <form
      onSubmit={[Function]}
    >
      <img
        alt="Visual of clouds"
        src="uploadimage.png"
      />
      <div>
        <input
          accept=".json"
          id="jsonFile"
          type="file"
        />
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
             Upload this file! 
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
